File: src/main/java/com/myapp/repository/UserRepository.java
package com.myapp.repository;

import com.myapp.domain.User;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.util.List;
import java.util.Optional;

public interface UserRepository extends JpaRepository<User, Long> {

    Optional<User> findOneByUsername(String username);

    @Query("select u from User u left join u.followedRelations f where f.follower = :user and u in :targetUsers")
    List<User> findFollowedBy(@Param("user") User user, @Param("targetUsers") List<User> targetUsers);

}

File: src/main/kotlin/com/myapp/repository/UserRepository.kt
package com.myapp.repository

import com.myapp.domain.User
import com.myapp.dto.page.Page

interface UserRepository {
    fun findOne(id: Long): User
    fun findOneWithStats(id: Long): User
    fun findOneByUsername(username: String): User?
    fun findAll(page: Int, size: Int): Page<User>
    fun create(user: User): User
    fun update(user: User)
}